go-controller/pkg/ovn/secondary_layer2_network_controller.go

```diff
@@ -0,0 +1,302 @@
+// could be called from a dummy Controller (only has CommonNetworkControllerInfo set)
+func (oc *SecondaryLayer2NetworkController) Cleanup(netName string) error {
+	klog.Infof("Delete OVN logical entities for %s network controller of network %s", types.Layer2Topology, netName)
+
+	// delete layer 2 logical switches
+	ops, err := libovsdbops.DeleteLogicalSwitchesWithPredicateOps(oc.nbClient, nil,
+		func(item *nbdb.LogicalSwitch) bool {
+			return item.ExternalIDs[types.NetworkExternalID] == netName
+		})
+	if err != nil {
+		return fmt.Errorf("failed to get ops for deleting switches of network %s: %v", netName, err)
+	}
+
+	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+	if err != nil {
+		return fmt.Errorf("failed to deleting switches of network %s: %v", netName, err)
+	}
+
+	return nil
+}
+
+func (oc *SecondaryLayer2NetworkController) Run() error {
+	klog.Infof("Starting all the Watchers for network %s ...", oc.GetNetworkName())
+	start := time.Now()
+
+	if err := oc.WatchPods(); err != nil {
+		return err
+	}
+
+	klog.Infof("Completing all the Watchers for network %s took %v", oc.GetNetworkName(), time.Since(start))
+
+	// controller is fully running and resource handlers have synced, update Topology version in OVN
+	if err := oc.updateL2TopologyVersion(); err != nil {
+		return fmt.Errorf("failed to update topology version for network %s: %v", oc.GetNetworkName(), err)
+	}
+
+	return nil
+}
+
+func (oc *SecondaryLayer2NetworkController) Init() error {
+	switchName := oc.GetNetworkScopedName(types.OVNLayer2Switch)
+	logicalSwitch := nbdb.LogicalSwitch{
+		Name:        switchName,
+		ExternalIDs: map[string]string{},
+	}
+	logicalSwitch.ExternalIDs[types.NetworkExternalID] = oc.GetNetworkName()
+	logicalSwitch.ExternalIDs[types.TopologyExternalID] = oc.TopologyType()
+
+	layer2NetConfInfo := oc.NetConfInfo.(*util.Layer2NetConfInfo)
+
+	hostSubnets := make([]*net.IPNet, 0, len(layer2NetConfInfo.ClusterSubnets))
+	for _, subnet := range layer2NetConfInfo.ClusterSubnets {
+		hostSubnets = append(hostSubnets, subnet)
+		if utilnet.IsIPv6CIDR(subnet) {
+			logicalSwitch.OtherConfig = map[string]string{"ipv6_prefix": subnet.IP.String()}
+		} else {
+			logicalSwitch.OtherConfig = map[string]string{"subnet": subnet.String()}
+		}
+	}
+
+	err := libovsdbops.CreateOrUpdateLogicalSwitch(oc.nbClient, &logicalSwitch, &logicalSwitch.OtherConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create logical switch %+v: %v", logicalSwitch, err)
+	}
+
+	err = libovsdbops.UpdateLogicalSwitchSetExternalIDs(oc.nbClient, &logicalSwitch)
+	if err != nil {
+		return fmt.Errorf("failed to update external-ids of logical switch %s, err: %v", switchName, err)
+	}
+
+	if err = oc.lsManager.AddSwitch(switchName, logicalSwitch.UUID, hostSubnets); err != nil {
+		return err
+	}
+
+	for _, excludeSubnet := range layer2NetConfInfo.ExcludeSubnets {
+		for excludeIP := excludeSubnet.IP; excludeSubnet.Contains(excludeIP); excludeIP = util.NextIP(excludeIP) {
```

Comment:
can you add a FIXME comment with a link to a github issue for this please? This is going to be really slow. We need to update the allocator to allow allocating a subnet, and do the required math there rather than iterating through every IP.

Reply:
I created an issue for that:
https://github.com/ovn-org/ovn-kubernetes/issues/3369